lib/puppet/provider/mount/parsed.rb

```diff
@@ -272,6 +272,49 @@ def self.mountinstances
     instances
   end
 
+  # Sort fstab entries so that
+  # a mount point is located inside another mount point appears later and
+  # a device that is located inside a mount point appears later.
+  #
+  # "Stable opportunistic insertion sort"
+  # Copy unsorted list A to sorted B by repeatedly
+  # * removing the first element from A
+  # * compare with each element in B, from first to last
+  # * insert before current element in B if smaller
+  # * insert at the end of B otherwise
+  # * finish if A is empty
+  # * repeat
+  #
+  # Stabilizes itself by inserting as late as possible.
+  #
+  # A record is "smaller" than another if
+  # * the mount point of A is a leading substring of B's mount point
+  # * the mount point of A is a leading substring of B's device (B is a bind mount)
+  #
+  # @note Does not really care about directory names and does substring
+  #   comparison only. So /media will be put before /mediastuff. This is
+  #   not necessary but will be rare and usually cause no harm, either.
+  #
+  # @api private
+  def self.order(records)
```

Comment:
As I feared: Passes some tests, but does break one of them.

Agree that fstabs are usually not large. Even if, this is an "at most once" kind of deal. The agent will do this only after all `mount` resources are synced and **only** if any of them changed.

I wasn't aware of the catch/throw penalty. If elegance costs performance, I will absolutely change to a flag as you suggested.